src/core/utils.ts

```diff
@@ -164,6 +166,25 @@ export function parseFilterArgs<
     : [arg1 || {}, arg2]) as [TFilters, TOptions]
 }
 
+export function mapQueryStatusFilter(
+  active?: boolean,
+  inactive?: boolean
+): QueryStatusFilter {
+  if (
+    (active === true && inactive === true) ||
+    (active == null && inactive == null)
+  ) {
+    return 'all'
+  } else if (active === false && inactive === false) {
+    return 'none'
+  } else {
+    // At this point, active|inactive can only be true|false or false|true
+    // so, when only one value is provided, the missing one has to be the negated value
+    const isActive = active ?? !inactive
+    return isActive ? 'active' : 'inactive'
+  }
+}
+
```

Comment:
@TkDodo here it is. This should cater for all the different permutations outlined in this test - https://github.com/tannerlinsley/react-query/pull/2275/files#diff-edc0f64b1261687e43194a6b6ec3881b7ece03d3d2c1236b2e087deae220fc82R311-R329

Reply:
wow, thanks a lot 🚀 

what I find interesting is that it seems that the actual fix only involved coercing the default value to `false` for `invalidateQueries` here:

https://github.com/tannerlinsley/react-query/pull/2275/files#diff-8e0c93e2ffd9983864be13d7b233dfc6d757e2fc95dd8c7fbc83a1a71aec6cbeR242

does that mean the issue was only present for `invalidateQueries`, but not for other methods like `refetchQueries` , because they don't have this mapping from `refetchActive -> active` and `refetchInactive -> inactive` ?

if so, can we maybe add another test especially for `invalidateQueries` - possibly the one that would fail before, but works now ? (The tests you've added on the queryClient are all with `refetchQueries` as far as I can see)

Reply:
Yes, that's one of the issues. `refetchInactive` should be defaulted to false.

Another issue is that the original code in `matchQuery` doesn't cater for scenario when both `active` and `inactive` are false, i.e. all queries shouldn't be matched.

Sure, I can add more test to cover for the failing scenario.

Reply:
Added more tests.

Actually this newly refactored QueryFilterStatus can be introduced as breaking change in the future. If we do, we will need to exclude `none` option as it doesn't make sense to run a filter function expecting nothing to be returned at all time.

Reply:
> In the interim, do you think we should bail out earlier from refetch when both active and inactive are false? Otherwise, it makes no sense to loop through this.queries.filter(query => matchQuery(filters, query)) when the result are always going to be []

while the assumption of having both active and inactive set to false _should_ yield empty array, it didn't do so with the previous implementation:

```
  let isActive

  if (inactive === false || (active && !inactive)) {
    isActive = true
  } else if (active === false || (inactive && !active)) {
    isActive = false
  }
```

here, `inactive` is checked first, and the `active` check was in an `else if`, so if you pass `{ inactive: false, active: false }`, you would still be matching all active queries 😅 

I still think this is a good change and can be seen as a fix 👍 . I would not do an early bailout because I consider this combination an invariant anyways. If we refactor this away from two boolean values later, we'll have less code to delete :)

Reply:
I'm referring to invalidateQueries, when both `refetchActive` and `refetchInactive` are set to `false` (likely as the result of setting `refetchActive` to false, because the other one is defaulted to false anyway).

This is how it looks like if we bails out from refetching,

```ts
// In invalidateQueries

return notifyManager.batch(() => {
  this.queryCache.findAll(filters).forEach(query => {
    query.invalidate()
  })
  
  const { refetchActive = true, refetchedInactive = false } = filters
  if(!refetchActive && !refetchInactive) {
    return;
  } else {
    const refetchFilters: QueryFilters = {
      ...filters,
      active: filters.refetchActive,
      inactive: filters.refetchInactive,
    }
    return this.refetchQueries(refetchFilters, options)
  }
})

```

Reply:
please see my other comment here: https://github.com/tannerlinsley/react-query/pull/2275#discussion_r659211920

I don't think this would work if we also choose to fallback to `filters.active` / `filters.inactive` for the `refetchFilters`.